store/store.go

```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
 		return err
 	}
 
-	has, err := s.Has(key)
+	// Get the key in primary storage
+	indexKey, err := s.index.Primary.IndexKey(key)
 	if err != nil {
 		return err
 	}
-	// we assume that a second put should NOT write twice
-	// the reason is we are assuming the key is the immutable hash for the value
-	if has {
-		return ErrKeyExists
-	}
-	fileOffset, err := s.index.Primary.Put(key, value)
+	// See if the key already exists and get offset
+	prevOffset, found, err := s.index.Get(indexKey)
 	if err != nil {
 		return err
 	}
-	indexKey, err := s.index.Primary.IndexKey(key)
+	// If found get the key and value stored in primary to see if it is the same
+	// (index only stores prefixes)
+	var storedKey []byte
+	var storedVal []byte
+	if found {
+		storedKey, storedVal, err = s.index.Primary.Get(prevOffset)
+		if err != nil {
+			return err
+		}
+	}
+
+	if bytes.Compare(key, storedKey)+bytes.Compare(value, storedVal) == 0 {
```

Comment:
This should be an && condition rather than sum so that you can skip the larger bytes.Compare if the first bytes.Compare is non zero. (Typical execution evaluates first expression in && and skips evaluating the second if the first fails. That's why people often write code like  `if a != nil && a.b == some value` -- the first check means the second check is safe to run without a nil pointer exception)

Reply:
*blocking*

Reply:
That is the result of me trusting the linter. I had it like that but I changed it 🤦‍♂️. Thanks!